app/models/concerns/profile_policy_association.rb

```diff
@@ -5,43 +5,18 @@ module ProfilePolicyAssociation
   extend ActiveSupport::Concern
 
   included do
-    after_destroy :destroy_policy_test_results
+    after_destroy :destroy_empty_policy
 
-    def policy
-      return self unless external
-
-      Profile.includes(:benchmark)
-             .older_than(created_at)
-             .find_by(account: account_id, external: false, ref_id: ref_id,
-                      benchmarks: { ref_id: benchmark.ref_id })
-    end
-
-    def policy_profiles
-      return Profile.none if account_id.nil?
-
-      Profile.includes(:benchmark)
-             .where(account: account_id, ref_id: ref_id,
-                    benchmarks: { ref_id: benchmark.ref_id })
-    end
-
-    def business_objective
-      BusinessObjective.find(business_objective_id) if business_objective_id
-    end
-
-    def business_objective_id
-      (policy || self).read_attribute(:business_objective_id)
-    end
+    belongs_to :policy, optional: true
+    delegate :business_objective, :business_objective_id, :update_hosts,
+             to: :policy, allow_nil: true
 
     def compliance_threshold
-      (policy || self).read_attribute(:compliance_threshold)
+      policy&.compliance_threshold || Policy::DEFAULT_COMPLIANCE_THRESHOLD
     end
 
-    def destroy_policy_test_results
-      if Settings.async
-        DestroyProfilesJob.perform_async(policy_profiles.pluck(:id))
-      else
-        DestroyProfilesJob.new.perform(policy_profiles.pluck(:id))
-      end
+    def destroy_empty_policy
+      policy.destroy if policy&.profiles&.empty?
```

Comment:
Observation: This could only happen if the current Profile is assigned to a policy and is not saved yet. Policy would have at least the current profile (`self`).

Reply:
(discard my previous comment)

Note, that this can resolve RHICOMPL-1031.